HttpRequestComponent::get_string - avoid copy

roitagar opened the pull request:
# What does this implement/fix? 

Complemeting @martgras's fix in #2987 - using `String` data type (same as returned by `HTTPClient::getString()`) and direct assignment to allow [Move assignment](https://en.cppreference.com/w/cpp/language/move_assignment) on ESP32.

On ESP8266 `HTTPClient::getString()` returns a `String&` to a member variable - therefore using a reference to avoid any kind of assignment/copy.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes esphome/issues#2903

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
http_request:
  id: my_req

interval:
  - interval: 5s
    then:
      - http_request.get:
          url: http://worldtimeapi.org/api/timezone/Asia/Jerusalem
          on_response:
            then:
              - logger.log:
                  format: "Got %s"
                  args: ['id(my_req).get_string()']
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

roitagar committed:
HttpRequestComponent::get_string - avoid copy

martgras reviewed (approved):
LGTM

oxan commented:
@martgras wrong button?

martgras commented:
Yes sorry

roitagar commented:
I just had another delayed enlightenment about this..
We can use `#ifdef` to make this a reference (`String&`) in ESP8266 and a `static String` in ESP32.
That should avoid copying AND be valid on both platforms..

I'm just not sure what are the flags to use, which should be the default and what other platforms/flavors are there out there / may pop out later on..
I also don't have an ESP8266 to test it with (but I think I can test just compilation without one, right?)

roitagar commented on esphome/components/http_request/http_request.cpp:
```diff
@@ -115,10 +115,10 @@ void HttpRequestComponent::close() {
-  // to keep a buffer alive.
-  static std::string str;
-  str = this->client_.getString().c_str();
+  // The static variable is here because HTTPClient::getString() returns a String on ESP32,
+  // and we need something to keep a buffer alive.
+  static String str;
+  str = this->client_.getString();
```
How about something like that?
```suggestion
#if defined (ESP32)
  // The static variable is here because HTTPClient::getString() returns a String on ESP32,
  // and we need something to keep a buffer alive.
  static String str;
#else
  // However on ESP8266, HTTPClient::getString() returns a String& to a member variable.
  // Leaving this the default so that any new platform either doesn't copy, or encounters a compilation error.
  String&
#endif
  str = this->client_.getString();
```

roitagar replied:
or maybe use `auto` in the `#else` clause, to allow additional return types like `std::string&`?
```suggestion
#if defined (ESP32)
  // The static variable is here because HTTPClient::getString() returns a String on ESP32,
  // and we need something to keep a buffer alive.
  static String str;
#else
  // However on ESP8266, HTTPClient::getString() returns a String& to a member variable.
  // Leaving this the default so that any new platform either doesn't copy, or encounters a compilation error.
  auto&
#endif
  str = this->client_.getString();
```

oxan replied:
We could, though the proper fix for this mess would be to change this whole method to return a `std::string` and transfer ownership to the caller.

roitagar replied:
Probably `String` and not `std::string`- to avoid copy.
I agree, that was my initial suggestion in esphome/issues#2903, but it breaks the API. Is that allowed?

Since the current comment of `HTTPClient::getString()` in both ESP32 and ESP8266 code says "`(may need lot of ram or trigger out of memory!)`", I'd hate to leave this code in a state that temporarily consumes twice the required amount..

martgras replied:
We try to avoid the Arduino String type in esphome.
Also if we manage to write an idf http component returning an Arduino String (reference) isn't optional.
So I agree it's not the optimal solution but I don't see better alternatives.

oxan replied:
Yeah, we don't use Arduino types in APIs in ESPHome. 

I think the copy could be avoided by implementing a `Stream` subclass that writes to a `std::string`, and using the `writeToStream` function, but honestly that effort is probably better invested in switching away from the Arduino HTTP libraries. They're not exactly high-quality, and we need to eventually do that for ESP-IDF anyway.

oxan replied:
What's the current state of this? I'm fine with the PR as-is or either suggestion, what do you prefer? Let's get this one done :)

roitagar replied:
Sorry, I was under the impression that we're waiting for your decision between the options.

Anyway, I've been experimenting a little more (and spotted a mistake in one of my suggestions - edited above already), pondering what's the right default (for a theoretical future platform which shares a compatible API but is not ESP32 nor ESP8266).

if a new platform/API returns by value: (like ESP32, possibly a different type though)
* `auto&` will cause a compilation error
* `auto` or `auto&&' will produce wrong behavior, and should output a compilation warning.
* `static String` will work correctly, but will use copy if it returns std::string

if a new platform/API returns a reference: (like ESP8266)
* `auto&` or `auto&&` works great and avoids copy
* `auto` will produce wrong behavior, and should output a compilation warning.
* `static String` will work correctly, but use copy

I also considered `static std::remove_reference_t<decltype(this->client_.getString())>` (which will use 'move' both with `String` and `std::string`, but the syntax looks too scary and I don't like that it repeats the call statement.

So it narrows down to deciding what is preferred on future incompatibility - compilation error vs. silent hidden copy. (former means `auto&`, latter means `static String`)
Initially I preferred the compilation error aproach, but thinking again at how much effort is required when adopting the code to a new platform, I thought maybe it's better to make the maintainer's life easier and potentially optimize later on.

What is the preferred spirit in this project? Tell me and I'll update the PR accordingly and test it (to be optimal for both ESP32 and ESP8266, and what you choose as default)

roitagar replied:
Sorry, I was under the impression that we're waiting for your decision between the options.

Anyway, I've been experimenting a little more (and spotted a mistake in one of my suggestions - edited above already), pondering what's the right default (for a theoretical future platform which shares a compatible API but is not ESP32 nor ESP8266).

I think I'll go with the `auto&` option for ESP8266 and default, I'll further test and prepare it soon.

I also considered `static std::remove_reference_t<decltype(this->client_.getString())>`, which will work correctly both with `String` and `std::string`.
I dislike it because (A) the syntax looks too scary, (B) it repeats the call statement, and (C) it will use copy on a theoretical future non-ESP8266 platform that returns a reference.

oxan replied:
It might very well have been blocked on me, I kinda lost track of this one :)

> I think I'll go with the `auto&` option for ESP8266 and default, I'll further test and prepare it soon.

Great!

roitagar committed:
HttpRequestComponent::get_string - avoid copy on ESP8266 too

- on ESP32 the temporary String object is move-assigned to a static variable
- on ESP8266 the reference to String is kept as reference

roitagar commented:
@oxan Done! Now it won't copy on ESP8266 too.
